### PR TITLE
Implementation of remaining metesat operations

### PR DESCRIPTION
**Pull request by koerberm**

### CODE REVIEW

**Comment by koerberm** [comment]
[MSG-Spezialoperatoren](https://trello.com/c/FBuL31eJ/220-msg-spezialoperatoren)

File: operators/src/processing/meteosat/temperature.rs [hunk 1]
```diff
@@ -0,0 +1,637 @@
...
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct TemperatureParams {
+    force_satellite: String,
+}
+
+impl Default for TemperatureParams {
+    fn default() -> Self {
+        TemperatureParams {
+            force_satellite: "".into(),
```
Comment (jdroenner) [inline operators/src/processing/meteosat/temperature.rs:38]: we should use a number (option)

File: operators/src/processing/meteosat/temperature.rs [hunk 2]
```diff
@@ -0,0 +1,637 @@
...
+                domain: Some("msg".into()),
+                key: "Satellite".into(),
+            },
+            channel_key: RasterPropertiesKey {
+                domain: Some("msg".into()),
+                key: "Channel".into(),
+            },
+            offset_key: RasterPropertiesKey {
+                domain: Some("msg".into()),
+                key: "CalibrationOffset".into(),
```
Comment (jdroenner) [inline operators/src/processing/meteosat/temperature.rs:189]: we can use offset and slope for this. the source can rename / map the properties.

**Review by jdroenner — CHANGES_REQUESTED** [review]
Looks good to me. There are some things we can change e.g. use enums.

File: operators/src/processing/meteosat/reflectance.rs [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+/// * `solar_correction` switch to enable solar correction.
+/// * `force_hrv` switch to force the use of the hrv channel.
+/// * `force_satellite` forces the use of the satellite with the given name.
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct ReflectanceParams {
+    pub solar_correction: bool,
+    #[serde(rename = "forceHRV")]
+    pub force_hrv: bool,
+    pub force_satellite: String,
```
[the remark this answers is not in the document]
Comment (koerberm) [reply operators/src/processing/meteosat/reflectance.rs]: I carried this over from mapping; Changed it to number now: 64b4dde20c007702dee325a438a5337d7c88f069

File: operators/src/processing/meteosat/reflectance.rs [hunk 4]
```diff
@@ -0,0 +1,664 @@
...
+                            let geos_coord = tile
+                                .global_geo_transform
+                                .grid_idx_to_center_coordinate_2d(idx);
+
+                            let (lat, lon) = channel.view_angle_lat_lon(geos_coord, 0.0);
+                            let (_, zenith) = sunpos.solar_azimuth_zenith(lat, lon);
+
+                            let idx = grid.shape.linear_space_index_unchecked(idx);
+                            let pixel = grid.data[idx];
+                            if !grid.is_no_data(pixel) {
```
[the remark this answers is not in the document]
Comment (koerberm) [reply operators/src/processing/meteosat/reflectance.rs:261]: OK: 14852ac08b3d0f8cd300b7c43a5027fac80f5e1e

File: operators/src/processing/meteosat/satellite.rs [hunk 5]
```diff
@@ -0,0 +1,636 @@
...
+    pub vc: f64,
+    pub alpha: f64,
+    pub beta: f64,
+}
+
+impl Satellite {
+    /// Retrieves the Meteosat satellite for the given `name`.
+    /// The name comparison is *case sensitive*.
+    pub fn satellite_by_name(name: &str) -> Result<&'static Satellite> {
+        match name {
```
[the remark this answers is not in the document]
Comment (koerberm) [reply operators/src/processing/meteosat/satellite.rs]: Method dropped, due to changing the forceSatellite parameter to number: 64b4dde20c007702dee325a438a5337d7c88f069

File: operators/src/processing/meteosat/reflectance.rs [hunk 6]
```diff
@@ -0,0 +1,664 @@
...
+            Satellite::satellite_by_name(self.params.force_satellite.as_str())
+        }
+    }
+
+    fn get_channel<'a>(
+        &self,
+        tile: &RasterTile2D<PixelOut>,
+        satellite: &'a Satellite,
+    ) -> Result<&'a Channel> {
+        if self.params.force_hrv {
```
[the remark this answers is not in the document]
Comment (koerberm) [reply operators/src/processing/meteosat/reflectance.rs:188]: I carried this over from mapping. Typically the channel is read from the tile's meta-data. Should we able to override the channel with a parameter? Then we should change it to a number (1-12)

File: operators/src/processing/meteosat/reflectance.rs [hunk 7]
```diff
@@ -0,0 +1,664 @@
...
+                    tile.time,
+                    tile.tile_position,
+                    tile.global_geo_transform,
+                    EmptyGrid::new(tile.grid_array.grid_shape(), OUT_NO_DATA_VALUE).into(),
+                    tile.properties,
+                )),
+                GridOrEmpty::Grid(grid) => {
+                    let satellite = self.get_satellite(&tile)?;
+                    let channel = self.get_channel(&tile, satellite)?;
+                    let timestamp = self.get_timestamp(&tile)?;
```
[the remark this answers is not in the document]
Comment (koerberm) [reply operators/src/processing/meteosat/reflectance.rs]: Ok: af5b54b9c06f912c02844c6f703b450a8ed018f5

File: operators/src/processing/meteosat/temperature.rs (hunk 1 above)
Comment (koerberm) [reply operators/src/processing/meteosat/temperature.rs]: See above: 64b4dde20c007702dee325a438a5337d7c88f069

File: operators/src/processing/meteosat/temperature.rs (hunk 2 above)
Comment (koerberm) [reply operators/src/processing/meteosat/temperature.rs:189]: I do not understand :D Simply change the key to  "offset"/"slope"?

Comment (koerberm) [reply operators/src/processing/meteosat/temperature.rs]: I added the mapping;

File: operators/src/processing/meteosat/temperature.rs [hunk 8]
```diff
@@ -0,0 +1,678 @@
...
+    }
+
+    // TODO: Cannot be tested, as the whole u8 range is a valid input. Think about making mock source u16
+    // #[tokio::test]
+    // async fn test_fail_illegal_input() {
+    //     let props = create_properties(Some(4), Some(1), Some(0.0), Some(1.0));
+    //     let mut params = TemperatureParams::default();
+    //     let res = process(props, params, None, Some(vec![1, 2, 3, 4, 1024, 0])).await;
+    //     assert!(res.is_err());
+    // }
```
Comment (ChristianBeilschmidt) [inline operators/src/processing/meteosat/temperature.rs:440]: Could we use a generic for the mock source? But maybe this is a lot of work?…

File: operators/src/util/sunpos.rs [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+            + day
+            - 32075;
+        let julian_date = aux2 as f64 - 0.5 + decimal_hours / 24.0;
+        (decimal_hours, julian_date - 2_451_545.0)
+    }
+
+    /// Calculate ecliptic coordinates (ecliptic longitude and obliquity of the
+    /// ecliptic in radians but without limiting the angle to be less than 2*Pi
+    /// (i.e., the result may be greater than 2*Pi)
+    fn eliptic_coordinates(elapsed_julian_days: f64) -> (f64, f64) {
```
Comment (ChristianBeilschmidt) [inline operators/src/util/sunpos.rs:87]:
```suggestion
    /// Calculate elliptic coordinates (elliptic longitude and obliquity of the
    /// elliptic in radians but without limiting the angle to be less than 2*Pi
    /// (i.e., the result may be greater than 2*Pi)
    fn elliptic_coordinates(elapsed_julian_days: f64) -> (f64, f64) {
```
if you mean *elliptic*

**Review by ChristianBeilschmidt — CHANGES_REQUESTED** [review]

File: operators/src/util/sunpos.rs (hunk 9 above)
Comment (koerberm) [reply operators/src/util/sunpos.rs:87]: I meant _ecliptic_ :D Will fix this.

Comment (koerberm) [reply operators/src/util/sunpos.rs]: 21be41da5f29bed84c76fa94581c5512e0eb0c8e

File: datatypes/src/raster/raster_properties.rs [hunk 10]
```diff
@@ -17,6 +18,32 @@ impl RasterProperties {
...
+                .get(key)
+                .ok_or(Error::MissingRasterProperty {
+                    property: key.to_string(),
+                })?
+                .clone(),
+        )?;
+        T::from_f64(val).ok_or(Error::WrongMetadataType)
+    }
+
+    pub fn get_string_property(&self, key: &RasterPropertiesKey) -> Result<String> {
```
[the remark this answers is not in the document]
Comment (koerberm) [reply datatypes/src/raster/raster_properties.rs]: 5982c6f14915e87c0fd1720571bd28a4b15a803c

File: datatypes/src/raster/raster_properties.rs [hunk 11]
```diff
@@ -69,6 +96,15 @@ impl TryFrom<RasterPropertiesEntry> for f64 {
...
 
+impl ToString for RasterPropertiesKey {
+    fn to_string(&self) -> String {
+        match &self.domain {
+            Some(prefix) => prefix.clone() + "." + &self.key,
+            None => self.key.clone(),
+        }
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (koerberm) [reply datatypes/src/raster/raster_properties.rs]: 26d161ba6cac8b4bebd63d24ea906bd94026efa1

File: operators/src/processing/meteosat/temperature.rs (hunk 8 above)
Comment (koerberm) [reply operators/src/processing/meteosat/temperature.rs:432]: Yes, it is some amount of work. Maybe do it as a separate task?

Comment (ChristianBeilschmidt) [reply operators/src/processing/meteosat/temperature.rs:432]: https://trello.com/c/xkolXr0t/340-mockrastersource-generischer-datentyp

**Review by ChristianBeilschmidt — APPROVED** [review]

**Pull request merged**